bpf/lib/policy.h

```diff
@@ -123,6 +123,37 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 localID,
 	}
 #endif /* ALLOW_ICMP_FRAG_NEEDED */
 
+	/* When the ICMP or ICMPv6 packet come, set its type to the key.dport*/
+	if (proto == IPPROTO_ICMP || proto==IPPROTO_ICMP_V6) {
+		void *data, *data_end;
+		__u8 icmp_type;
+		__u32 off;
+
+		if (proto==IPPROTO_ICMP){
+			struct iphdr *ip4;
+
+			if (!revalidate_data(ctx, &data, &data_end, &ip4))
+				return DROP_INVALID;
+			off = ((void *)ip4 - data) + ipv4_hdrlen(ip4);
+		}else{
+			struct ipv6hdr *ip6;
+
+			if (!revalidate_data(ctx, &data, &data_end, &ip6))
+				return DROP_INVALID;
+			off = ((void *)ip6 - data) + ipv6_hdrlen(ctx, ETH_HLEN, &ip6->nexthdr);
+		}
```

Comment:
Doesn't it work if you just use `struct iphdr` for both ICMP and ICMPv6 (with a comment to explain they have the type at the same offset)?

Reply:
I tried, but it didn't work.
I think the length of ip4 and ip6 is different, so the value of `off` is different.

Reply:
Ah, right. You still need the two structures for `ip6->nexthdr` :disappointed: 
You could still move the `revalidate_data` out but I'm not sure that will buy us a lot w.r.t. the complexity. The compiler should be able to do the same.